File: app/src/main/java/com/metahumanz/pacilread/stats/annual/AnnualReportMetric.java
package com.metahumanz.pacilread.stats.annual;

import com.metahumanz.pacilread.stats.ReadingStatsUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

public enum AnnualReportMetric {
    TOTAL_DURATION("total_duration"),
    TOTAL_CHARS("total_chars"),
    READING_BOOKS("reading_books"),
    READING_DAYS("reading_days"),
    LONGEST_STREAK("longest_streak"),
    FINISHED_BOOKS("finished_books"),
    TOP_BOOK("top_book"),
    TOP_AUTHOR("top_author"),
    TOP_TAG("top_tag"),
    TOP_SERIES("top_series"),
    PEAK_MONTH("peak_month"),
    ACTIVE_MONTHS("active_months"),
    DAILY_AVERAGE("daily_average"),
    BOOK_STATUS("book_status"),
    BOOK_SPEED("book_speed");

    public final String key;

    AnnualReportMetric(String key) {
        this.key = key;
    }

    public String label(AnnualReportData report) {
        boolean bookScope = report != null && report.isBookScope();
        switch (this) {
            case TOTAL_DURATION:
                return bookScope ? "本书时长" : "阅读总时长";
            case TOTAL_CHARS:
                return bookScope ? "本书字数" : "阅读字数";
            case READING_BOOKS:
                return "阅读书籍";
            case READING_DAYS:
                return "阅读天数";
            case LONGEST_STREAK:
                return "最长连续";
            case FINISHED_BOOKS:
                return "完成书籍";
            case TOP_BOOK:
                return "Top 书籍";
            case TOP_AUTHOR:
                return bookScope ? "作者" : "常读作者";
            case TOP_TAG:
                return bookScope ? "标签" : "常读标签";
            case TOP_SERIES:
                return bookScope ? "系列" : "常读系列";
            case PEAK_MONTH:
                return report != null && report.isYearReport() ? "最活跃月份" : "最活跃日";
            case ACTIVE_MONTHS:
                return report != null && report.isYearReport() ? "活跃月份" : "活跃天数";
            case DAILY_AVERAGE:
                return "日均阅读";
            case BOOK_STATUS:
                return "完成状态";
            case BOOK_SPEED:
                return "阅读速度";
            default:
                return "";
        }
    }

    public String value(AnnualReportData report) {
        if (report == null) {
            return "";
        }
        switch (this) {
            case TOTAL_DURATION:
                return ReadingStatsUtils.formatDuration(report.totalSeconds);
            case TOTAL_CHARS:
                return formatNumber(report.totalChars) + " 字";
            case READING_BOOKS:
                return report.readingBooks + " 本";
            case READING_DAYS:
                return report.readingDays + " 天";
            case LONGEST_STREAK:
                return report.longestStreak + " 天";
            case FINISHED_BOOKS:
                return report.finishedBooks + " 本";
            case TOP_BOOK:
                return clean(report.topBook);
            case TOP_AUTHOR:
                return report.isBookScope() ? clean(report.bookAuthor) : clean(report.topAuthor);
            case TOP_TAG:
                return clean(report.topTag);
            case TOP_SERIES:
                return clean(report.topSeries);
            case PEAK_MONTH:
                return report.isYearReport()
                        ? (report.peakMonth > 0 ? report.peakMonth + " 月" : "")
                        : clean(report.peakRhythmLabel);
            case ACTIVE_MONTHS:
                int active = report.isYearReport() ? report.activeMonths : report.activeRhythmSlots;
                String unit = report.isYearReport() ? "个月" : "天";
                return active + " " + unit;
            case DAILY_AVERAGE:
                return ReadingStatsUtils.formatDuration(report.averageSecondsPerReadingDay());
            case BOOK_STATUS:
                return clean(report.statusText);
            case BOOK_SPEED:
                return report.readingSpeedCharsPerMinute > 0
                        ? report.readingSpeedCharsPerMinute + " 字/分"
                        : "";
            default:
                return "";
        }
    }

    public boolean supports(AnnualReportData report) {
        if (report == null) {
            return false;
        }
        boolean bookScope = report.isBookScope();
        switch (this) {
            case FINISHED_BOOKS:
                return !bookScope && report.isYearReport();
            case READING_BOOKS:
            case TOP_BOOK:
            case DAILY_AVERAGE:
                return !bookScope;
            case BOOK_STATUS:
            case BOOK_SPEED:
                return bookScope;
            default:
                return true;
        }
    }

    public boolean isAvailable(AnnualReportData report) {
        if (!supports(report)) {
            return false;
        }
        switch (this) {
            case TOTAL_DURATION:
                return report.totalSeconds > 0;
            case TOTAL_CHARS:
                return report.totalChars > 0;
            case READING_BOOKS:
                return report.readingBooks > 0;
            case READING_DAYS:
                return report.readingDays > 0;
            case LONGEST_STREAK:
                return report.longestStreak > 0;
            case FINISHED_BOOKS:
                return true;
            case TOP_BOOK:
                return hasText(report.topBook);
            case TOP_AUTHOR:
                return hasText(report.isBookScope() ? report.bookAuthor : report.topAuthor);
            case TOP_TAG:
                return hasText(report.topTag);
            case TOP_SERIES:
                return hasText(report.topSeries);
            case PEAK_MONTH:
                return report.isYearReport() ? report.peakMonth > 0 : hasText(report.peakRhythmLabel);
            case ACTIVE_MONTHS:
                return (report.isYearReport() ? report.activeMonths : report.activeRhythmSlots) > 0;
            case DAILY_AVERAGE:
                return report.averageSecondsPerReadingDay() > 0;
            case BOOK_STATUS:
                return hasText(report.statusText);
            case BOOK_SPEED:
                return report.readingSpeedCharsPerMinute > 0;
            default:
                return false;
        }
    }

    public static AnnualReportMetric fromKey(String key) {
        if (key == null) {
            return null;
        }
        String safeKey = key.trim();
        for (AnnualReportMetric metric : values()) {
            if (metric.key.equals(safeKey)) {
                return metric;
            }
        }
        return null;
    }

    public static List<String> parseKeys(String serialized) {
        List<String> keys = new ArrayList<>();
        if (serialized == null || serialized.trim().isEmpty()) {
            return keys;
        }
        String[] parts = serialized.split(",");
        for (String part : parts) {
            String key = part == null ? "" : part.trim();
            if (!key.isEmpty()) {
                keys.add(key);
            }
        }
        return keys;
    }

    public static String serialize(List<AnnualReportMetric> metrics) {
        if (metrics == null || metrics.isEmpty()) {
            return "";
        }
        StringBuilder builder = new StringBuilder();
        for (AnnualReportMetric metric : metrics) {
            if (metric == null) {
                continue;
            }
            if (builder.length() > 0) {
                builder.append(',');
            }
            builder.append(metric.key);
        }
        return builder.toString();
    }

    public static List<AnnualReportMetric> selectionFromKeys(AnnualReportData report, List<String> keys) {
        List<AnnualReportMetric> selected = new ArrayList<>();
        if (keys != null) {
            for (String key : keys) {
                AnnualReportMetric metric = fromKey(key);
                if (metric != null && !selected.contains(metric)) {
                    selected.add(metric);
                }
            }
        }
        return sanitizeMetrics(report, selected);
    }

    public static List<AnnualReportMetric> sanitizeMetrics(AnnualReportData report, List<AnnualReportMetric> metrics) {
        List<AnnualReportMetric> available = availableFor(report);
        List<AnnualReportMetric> selected = new ArrayList<>();
        if (metrics != null) {
            for (AnnualReportMetric metric : metrics) {
                if (metric != null && available.contains(metric) && !selected.contains(metric)) {
                    selected.add(metric);
                }
                if (selected.size() == 3) {
                    return selected;
                }
            }
        }
        for (AnnualReportMetric metric : defaultOrder(report)) {
            if (available.contains(metric) && !selected.contains(metric)) {
                selected.add(metric);
            }
            if (selected.size() == 3) {
                return selected;
            }
        }
        for (AnnualReportMetric metric : available) {
            if (!selected.contains(metric)) {
                selected.add(metric);
            }
            if (selected.size() == 3) {
                break;
            }
        }
        return selected;
    }

    public static List<AnnualReportMetric> availableFor(AnnualReportData report) {
        List<AnnualReportMetric> available = new ArrayList<>();
        if (report == null) {
            return available;
        }
        AnnualReportMetric[] order = report.isBookScope() ? bookOrder() : globalOrder();
        for (AnnualReportMetric metric : order) {
            if (metric.isAvailable(report) && !available.contains(metric)) {
                available.add(metric);
            }
        }
        for (AnnualReportMetric metric : defaultOrder(report)) {
            if (metric.supports(report) && !available.contains(metric)) {
                available.add(metric);
            }
            if (available.size() >= 3) {
                break;
            }
        }
        return available;
    }

    private static AnnualReportMetric[] defaultOrder(AnnualReportData report) {
        if (report != null && report.isBookScope()) {
            return new AnnualReportMetric[]{
                    READING_DAYS,
                    LONGEST_STREAK,
                    BOOK_STATUS,
                    TOTAL_DURATION,
                    TOTAL_CHARS
            };
        }
        return new AnnualReportMetric[]{
                TOTAL_DURATION,
                TOTAL_CHARS,
                report != null && report.isYearReport() ? FINISHED_BOOKS : READING_BOOKS,
                READING_DAYS,
                LONGEST_STREAK
        };
    }

    private static AnnualReportMetric[] globalOrder() {
        return new AnnualReportMetric[]{
                TOTAL_DURATION,
                TOTAL_CHARS,
                READING_BOOKS,
                READING_DAYS,
                LONGEST_STREAK,
                FINISHED_BOOKS,
                TOP_BOOK,
                TOP_AUTHOR,
                TOP_TAG,
                TOP_SERIES,
                PEAK_MONTH,
                ACTIVE_MONTHS,
                DAILY_AVERAGE
        };
    }

    private static AnnualReportMetric[] bookOrder() {
        return new AnnualReportMetric[]{
                TOTAL_DURATION,
                TOTAL_CHARS,
                READING_DAYS,
                LONGEST_STREAK,
                BOOK_STATUS,
                BOOK_SPEED,
                PEAK_MONTH,
                ACTIVE_MONTHS,
                TOP_AUTHOR,
                TOP_TAG,
                TOP_SERIES
        };
    }

    private static String formatNumber(int value) {
        return String.format(Locale.SIMPLIFIED_CHINESE, "%,d", Math.max(value, 0));
    }

    private static boolean hasText(String value) {
        return value != null && !value.trim().isEmpty();
    }

    private static String clean(String value) {
        return value == null ? "" : value.trim();
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/stats/annual/AnnualReportMetric.kt
package com.metahumanz.pacilread.stats.annual

import com.metahumanz.pacilread.stats.ReadingStatsUtils
import java.util.Locale

enum class AnnualReportMetric(@JvmField val key: String) {
    TOTAL_DURATION("total_duration"),
    TOTAL_CHARS("total_chars"),
    READING_BOOKS("reading_books"),
    READING_DAYS("reading_days"),
    LONGEST_STREAK("longest_streak"),
    FINISHED_BOOKS("finished_books"),
    TOP_BOOK("top_book"),
    TOP_AUTHOR("top_author"),
    TOP_TAG("top_tag"),
    TOP_SERIES("top_series"),
    PEAK_MONTH("peak_month"),
    ACTIVE_MONTHS("active_months"),
    DAILY_AVERAGE("daily_average"),
    BOOK_STATUS("book_status"),
    BOOK_SPEED("book_speed");

    fun label(report: AnnualReportData?): String {
        val bookScope = report?.isBookScope() == true
        return when (this) {
            TOTAL_DURATION -> if (bookScope) "本书时长" else "阅读总时长"
            TOTAL_CHARS -> if (bookScope) "本书字数" else "阅读字数"
            READING_BOOKS -> "阅读书籍"
            READING_DAYS -> "阅读天数"
            LONGEST_STREAK -> "最长连续"
            FINISHED_BOOKS -> "完成书籍"
            TOP_BOOK -> "Top 书籍"
            TOP_AUTHOR -> if (bookScope) "作者" else "常读作者"
            TOP_TAG -> if (bookScope) "标签" else "常读标签"
            TOP_SERIES -> if (bookScope) "系列" else "常读系列"
            PEAK_MONTH -> if (report?.isYearReport() == true) "最活跃月份" else "最活跃日"
            ACTIVE_MONTHS -> if (report?.isYearReport() == true) "活跃月份" else "活跃天数"
            DAILY_AVERAGE -> "日均阅读"
            BOOK_STATUS -> "完成状态"
            BOOK_SPEED -> "阅读速度"
        }
    }

    fun value(report: AnnualReportData?): String {
        if (report == null) return ""
        return when (this) {
            TOTAL_DURATION -> ReadingStatsUtils.formatDuration(report.totalSeconds)
            TOTAL_CHARS -> "${formatNumber(report.totalChars)} 字"
            READING_BOOKS -> "${report.readingBooks} 本"
            READING_DAYS -> "${report.readingDays} 天"
            LONGEST_STREAK -> "${report.longestStreak} 天"
            FINISHED_BOOKS -> "${report.finishedBooks} 本"
            TOP_BOOK -> clean(report.topBook)
            TOP_AUTHOR -> clean(if (report.isBookScope()) report.bookAuthor else report.topAuthor)
            TOP_TAG -> clean(report.topTag)
            TOP_SERIES -> clean(report.topSeries)
            PEAK_MONTH -> if (report.isYearReport()) {
                if (report.isLast365DaysReport()) clean(report.peakRhythmLabel)
                else if (report.peakMonth > 0) "${report.peakMonth} 月" else ""
            } else clean(report.peakRhythmLabel)
            ACTIVE_MONTHS -> {
                val active = if (report.isYearReport()) report.activeMonths else report.activeRhythmSlots
                "$active ${if (report.isYearReport()) "个月" else "天"}"
            }
            DAILY_AVERAGE -> ReadingStatsUtils.formatDuration(report.averageSecondsPerReadingDay())
            BOOK_STATUS -> clean(report.statusText)
            BOOK_SPEED -> if (report.readingSpeedCharsPerMinute > 0) "${report.readingSpeedCharsPerMinute} 字/分" else ""
        }
    }

    fun supports(report: AnnualReportData?): Boolean {
        if (report == null) return false
        val bookScope = report.isBookScope()
        return when (this) {
            FINISHED_BOOKS -> !bookScope && report.isYearReport()
            READING_BOOKS, TOP_BOOK, DAILY_AVERAGE -> !bookScope
            BOOK_STATUS, BOOK_SPEED -> bookScope
            else -> true
        }
    }

    fun isAvailable(report: AnnualReportData?): Boolean {
        if (!supports(report) || report == null) return false
        return when (this) {
            TOTAL_DURATION -> report.totalSeconds > 0
            TOTAL_CHARS -> report.totalChars > 0
            READING_BOOKS -> report.readingBooks > 0
            READING_DAYS -> report.readingDays > 0
            LONGEST_STREAK -> report.longestStreak > 0
            FINISHED_BOOKS -> true
            TOP_BOOK -> hasText(report.topBook)
            TOP_AUTHOR -> hasText(if (report.isBookScope()) report.bookAuthor else report.topAuthor)
            TOP_TAG -> hasText(report.topTag)
            TOP_SERIES -> hasText(report.topSeries)
            PEAK_MONTH -> if (report.isYearReport()) {
                if (report.isLast365DaysReport()) hasText(report.peakRhythmLabel) else report.peakMonth > 0
            } else hasText(report.peakRhythmLabel)
            ACTIVE_MONTHS -> (if (report.isYearReport()) report.activeMonths else report.activeRhythmSlots) > 0
            DAILY_AVERAGE -> report.averageSecondsPerReadingDay() > 0
            BOOK_STATUS -> hasText(report.statusText)
            BOOK_SPEED -> report.readingSpeedCharsPerMinute > 0
        }
    }

    companion object {
        @JvmStatic
        fun fromKey(key: String?): AnnualReportMetric? {
            if (key == null) return null
            val safeKey = key.trim()
            for (metric in entries) if (metric.key == safeKey) return metric
            return null
        }

        @JvmStatic
        fun parseKeys(serialized: String?): List<String> {
            val keys = ArrayList<String>()
            if (serialized.isNullOrBlank()) return keys
            for (part in serialized.split(',')) part.trim().takeIf { it.isNotEmpty() }?.let(keys::add)
            return keys
        }

        @JvmStatic
        fun serialize(metrics: List<AnnualReportMetric?>?): String {
            if (metrics.isNullOrEmpty()) return ""
            val builder = StringBuilder()
            for (metric in metrics) {
                if (metric == null) continue
                if (builder.isNotEmpty()) builder.append(',')
                builder.append(metric.key)
            }
            return builder.toString()
        }

        @JvmStatic
        fun selectionFromKeys(report: AnnualReportData?, keys: List<String>?): List<AnnualReportMetric> {
            val selected = ArrayList<AnnualReportMetric>()
            if (keys != null) for (key in keys) fromKey(key)?.let { if (!selected.contains(it)) selected.add(it) }
            return sanitizeMetrics(report, selected)
        }

        @JvmStatic
        fun sanitizeMetrics(report: AnnualReportData?, metrics: List<AnnualReportMetric?>?): List<AnnualReportMetric> {
            val available = availableFor(report)
            val selected = ArrayList<AnnualReportMetric>()
            if (metrics != null) for (metric in metrics) {
                if (metric != null && available.contains(metric) && !selected.contains(metric)) selected.add(metric)
                if (selected.size == 3) return selected
            }
            for (metric in defaultOrder(report)) {
                if (available.contains(metric) && !selected.contains(metric)) selected.add(metric)
                if (selected.size == 3) return selected
            }
            for (metric in available) {
                if (!selected.contains(metric)) selected.add(metric)
                if (selected.size == 3) break
            }
            return selected
        }

        @JvmStatic
        fun availableFor(report: AnnualReportData?): List<AnnualReportMetric> {
            val available = ArrayList<AnnualReportMetric>()
            if (report == null) return available
            for (metric in if (report.isBookScope()) bookOrder() else globalOrder()) {
                if (metric.isAvailable(report) && !available.contains(metric)) available.add(metric)
            }
            for (metric in defaultOrder(report)) {
                if (metric.supports(report) && !available.contains(metric)) available.add(metric)
                if (available.size >= 3) break
            }
            return available
        }

        private fun defaultOrder(report: AnnualReportData?): Array<AnnualReportMetric> = if (report?.isBookScope() == true) {
            arrayOf(READING_DAYS, LONGEST_STREAK, BOOK_STATUS, TOTAL_DURATION, TOTAL_CHARS)
        } else {
            arrayOf(READING_DAYS, LONGEST_STREAK, if (report?.isYearReport() == true) FINISHED_BOOKS else READING_BOOKS, TOTAL_DURATION, TOTAL_CHARS)
        }

        private fun globalOrder() = arrayOf(
            TOTAL_DURATION, TOTAL_CHARS, READING_BOOKS, READING_DAYS, LONGEST_STREAK, FINISHED_BOOKS,
            TOP_BOOK, TOP_AUTHOR, TOP_TAG, TOP_SERIES, PEAK_MONTH, ACTIVE_MONTHS, DAILY_AVERAGE,
        )

        private fun bookOrder() = arrayOf(
            TOTAL_DURATION, TOTAL_CHARS, READING_DAYS, LONGEST_STREAK, BOOK_STATUS, BOOK_SPEED,
            PEAK_MONTH, ACTIVE_MONTHS, TOP_AUTHOR, TOP_TAG, TOP_SERIES,
        )

        private fun formatNumber(value: Int): String = String.format(Locale.SIMPLIFIED_CHINESE, "%,d", Math.max(value, 0))
        private fun hasText(value: String?): Boolean = !value.isNullOrBlank()
        private fun clean(value: String?): String = value?.trim().orEmpty()
    }
}
